internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go

```diff
@@ -505,6 +505,147 @@ func (s *Server) List(req *machine.ListRequest, obj machine.MachineService_ListS
+		if path == "" {
+			path = "/"
+		}
+
+		_, err := os.Stat(path)
+		if err == os.ErrNotExist {
+			return fmt.Errorf("%s no such file or directory", path)
+		}
+
+		files, err := archiver.Walker(obj.Context(), path, archiver.WithMaxRecurseDepth(-1))
+		if err != nil {
+			return err
+		}
+
+		folders := map[string]*machine.DiskUsageInfo{}
+
+		// send a record back to client if the message shouldn't be skipped
+		// at the same time use record information for folder size estimation
+		sendSize := func(info *machine.DiskUsageInfo, depth int32, isDir bool) {
+			prefix := strings.TrimRight(filepath.Dir(info.Name), "/")
+			if folder, ok := folders[prefix]; ok {
+				folder.Size += info.Size
+			}
+
+			// recursion depth check
+			skip := depth >= req.RecursionDepth && req.RecursionDepth > 0
+			// skip files check
+			skip = skip || !isDir && !req.All
+			// threshold check
+			skip = skip || req.Threshold > 0 && info.Size < req.Threshold
+			skip = skip || req.Threshold < 0 && info.Size > -req.Threshold
+
+			if skip {
+				return
+			}
+
+			err = obj.Send(info)
+		}
+
+		var (
+			depth     int32
+			prefix    = path
+			rootDepth = int32(strings.Count(path, archiver.OSPathSeparator))
+		)
+
+		// flush all folder sizes until we get to the common prefix
+		flushFolders := func(prefix, nextPrefix string) {
+			for !strings.HasPrefix(nextPrefix, prefix) {
+				currentDepth := int32(strings.Count(prefix, archiver.OSPathSeparator)) - rootDepth
+
+				if folder, ok := folders[prefix]; ok {
+					sendSize(folder, currentDepth, true)
+				}
+
+				prefix = strings.TrimRight(filepath.Dir(prefix), "/")
+			}
+		}
+
+		for fi := range files {
+			if fi.Error != nil {
+				err = obj.Send(
+					&machine.DiskUsageInfo{
+						Name:         fi.FullPath,
+						RelativeName: fi.RelPath,
+						Error:        fi.Error.Error(),
+					},
+				)
+			} else {
+				currentDepth := int32(strings.Count(fi.FullPath, archiver.OSPathSeparator)) - rootDepth
+				size := fi.FileInfo.Size()
+				if size < 0 {
+					size = 0
+				}
+
+				// kcore file size gives wrong value, this code should be smarter when it reads it
+				// TODO: figure out better way to skip such file
+				if fi.FullPath == "/proc/kcore" {
+					size = 0
+				}
+
+				if fi.FileInfo.IsDir() {
+					folders[strings.TrimRight(fi.FullPath, "/")] = &machine.DiskUsageInfo{
+						Name:         fi.FullPath,
+						RelativeName: fi.RelPath,
+						Size:         size,
+					}
+				} else {
+					sendSize(&machine.DiskUsageInfo{
+						Name:         fi.FullPath,
+						RelativeName: fi.RelPath,
+						Size:         size,
+					}, currentDepth, false)
+				}
+
+				// depth goes down when walker gets to the next sibling folder
+				if currentDepth < depth {
+					nextPrefix := fi.FullPath
+					flushFolders(prefix, nextPrefix)
+					prefix = nextPrefix
+				}
+
+				if fi.FileInfo.IsDir() {
+					prefix = fi.FullPath
+				}
+				depth = currentDepth
+			}
+
+			if err != nil {
```

Comment:
should this be moved closer to `Send()` above, as it's the only error as it seems?